### PR TITLE
hashable instrument/sample/detectors objects

### PR DESCRIPTION
**Pull request by abuts**
Solves Re #1790. 

Should be dealt wich after Re #1789 is merged.

All objects targeted to be in the conteners have been made hashable. Bunch of unit tests written validating that hashable works. 
Tests allowed to fix number of errors (not only with hashable).

This should resolve the issue with performance of large number of small cuts, chapter 2 in epic and hugely simplify code to deal with hashes.

Profiling using small script below, similar to one used by the complainer show:
```
30 small cuts before hashing implemented							: 174sec
30 small cuts with hashing and data placed on disk(quck `upgrade_file_format` run on files v4)	:  65sec.
```
There is still 2-times slow down between new and old Horace, but this slow down should be addressed by https://github.com/pace-neutrons/Horace/issues/896
as `mex` was providing approximately 2-fold acceleration for slow operations below, when invoked.

No traces of hashing recalculation in new code timing, though some code blocks seems may be optimized more:

OldVsNewPerformance

Testing code:
```MATLAB
proj = line_proj;
file = 'e:\SHARE\Fe\Data\sqw\sqw2024\Fe_ei401_align_tc.sqw';
profile on
nsteps = 30;
times = zeros(1,nsteps);

for i=1:nsteps 
    e0 = (i-1)*10;    
    t0 = tic;    
    w2= cut(file,proj,0.04,0.04,[-0.1,0.1],[e0-10,e0+10],'-nopix');
    times(i) = toc(t0);
    %plot(w2);
    lz(0,5-(i-1)*(4.99/(nsteps-1)));
    %pause(1)
end
fprintf("Total time %dsec\n",sum(times));
profile viewer
```

### CODE REVIEW

File: _test/common_functions/hashable_obj_tester.m [hunk 1]
```diff
@@ -0,0 +1,56 @@
+function hashable_obj_tester(hobj,values,names)
+%hashable_obj_tester Generic function which validates correct hashable
+%object operations.
+%
+% Checks if all hashable properties invalidate hash, hash calculations work
+% and hash is stored-restored correctly
+%
+% Inputs:
+% hobj  -- instance of hashable object, better with proper values set
```
Comment (cmarooney-stfc) [inline _test/common_functions/hashable_obj_tester.m:9]: unclear what is meant by better. What will happen if they aren't all set

File: _test/common_functions/hashable_obj_tester.m [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    hobj = hobj.build_hash();
+end
+
+for i=1:n_flds
+    assertTrue(hobj.hash_defined)
+    hobj.(flds{i}) = values{i};
+    if hobj.hash_defined
+        assertFalse(true, ...
+            sprintf('*** Hash remains defined for changed property: %s ',flds{i}));
+    end
```
Comment (cmarooney-stfc) [inline _test/common_functions/hashable_obj_tester.m:41]: unclear what this is doing. assertFalse(true...) looks strange

File: _test/common_functions/hashable_obj_tester.m [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+S = hobj.to_struct();
+rec_obj= hobj.from_struct(S);
+
+assertTrue(hobj.hash_defined);
+assertTrue(hobj == rec_obj)
+
+obj_arr = [hobj,hobj];
+S = obj_arr.to_struct();
+rec_arr= hobj.from_struct(S);
```
Comment (cmarooney-stfc) [inline _test/common_functions/hashable_obj_tester.m:53]: unclear what this is doing

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]

File: herbert_core/utilities/classes/@hashable/private/to_hashable_array_.m [hunk 4]
```diff
@@ -32,12 +32,16 @@
...
+                % part of constructor we do not need to check combo_values.
+                % If may fail.
                 obj(j).(field_name) = val;
+                obj(j).do_check_combo_arg = true;                
             end
             if iscell(hash) % use array of hashes as source for final hash
                 hash = strjoin(hash ,'');
             end
-            tm = typecast(hash,'uint8');
+            tm = uint8(hash); % beware of possibility hash changing type in a future
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/utilities/classes/@hashable/private/to_hashable_array_.m:44]: it is subject for discussion. If we use 64 bit hash instead of 128bit hash, container search/modifications operation become an order of magnitude faster. 

May be it worth switching to such hash. If hash is uint64, this statement needs to be modified. Tests will show that there is a bug with hashes, but probably in different place than the row 44, so the comment is intended for debugging in this situation.

File: herbert_core/classes/instrument_classes/instrument/instruments/@IX_inst_DGfermi/IX_inst_DGfermi.m [hunk 5]
```diff
@@ -70,6 +70,7 @@
         function obj=set.energy(obj,val)
             obj.moderator_.energy = val;
             obj.fermi_chopper_.energy = val;
+            obj = obj.clear_hash();
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/instruments/@IX_inst_DGfermi/IX_inst_DGfermi.m:73]: I am not aware of way of doing this in Matlab. In Python it could probably be done through wrapper functions. 
In any situation, it would be some kind of global setter function used everywhere and overloaded in `hashable`  We do not have such function, so I am taking approach choosen in `serializable` for `check_common_args`

Subclasses should use common methods of parent class.  This is why parent class protected methods are for. And public too, though public may be used for other purposes too.

File: herbert_core/classes/instrument_classes/instrument/instruments/@IX_inst_DGdisk/IX_inst_DGdisk.m [hunk 6]
```diff
@@ -235,6 +235,7 @@
             mcm = obj.mod_shape_mono;
             mcm.do_check_combo_arg = true;
             obj.mod_shape_mono_ = mcm.check_combo_arg();
+            obj = obj.clear_hash();
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/instruments/@IX_inst_DGdisk/IX_inst_DGdisk.m:238]: not as I am aware of.

File: herbert_core/classes/instrument_classes/instrument/instruments/@IX_inst/IX_inst.m [hunk 7]
```diff
@@ -106,6 +106,7 @@
                     'The source name must be a character string')
             end
             obj.name_ = char(val);
+            obj = obj.clear_hash();
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/instruments/@IX_inst/IX_inst.m:109]: ibid. It is ~300 such insertions in the code.

File: herbert_core/classes/instrument_classes/instrument/detectors/@IX_detector_array/IX_detector_array.m [hunk 8]
```diff
@@ -553,13 +553,17 @@
...
-            val.height   = obj.det_bank_(1).height; 
+            val.width    = obj.det_bank_(1).width;
+            val.height   = obj.det_bank_(1).height;
             val.filename = obj.filename_;
-            val.filepath = obj.filepath_;        
+            val.filepath = obj.filepath_;
+        end
+        function obj = check_combo_arg(obj)
+            % TODO: at least array equal length should be validated
+            obj = obj.clear_hash();
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/detectors/@IX_detector_array/IX_detector_array.m:563]: This is the qestion for the `IX_detector_array` developer. This is where I put this question to him. I am not aware of `IX_detector_array`  purpose and why its arguments are not beeing validated properly. May be it is work in progress.

[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/detectors/@IX_detector_array/IX_detector_array.m:563]: because checking is called in setters. Assuming that settings have changed, hash have to be invalidated.

File: herbert_core/classes/instrument_classes/instrument/detectors/@IX_det_He3tube/IX_det_He3tube.m [hunk 9]
```diff
@@ -192,17 +192,18 @@
...
+
             % Check the consistency of tube diameter and wall thickness
             if  ~all(obj.dia_>=2*obj.wall_)
                 error('HERBERT:IX_det_He3tube:invalid_argument',...
                     ['Tube diameter(s) must be greater or equal to twice ',...
                     'the wall thickness(es)'])
             end
-        end
-        
+            obj = obj.clear_hash();
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/detectors/@IX_det_He3tube/IX_det_He3tube.m:202]: see above

File: herbert_core/classes/instrument_classes/instrument/@IX_doubledisk_chopper/IX_doubledisk_chopper.m [hunk 10]
```diff
@@ -289,6 +292,7 @@
             end
             if do_recompute_pdf
                 obj.pdf_ = recompute_pdf_(obj);   % recompute the lookup table
+                obj = obj.clear_hash();
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/@IX_doubledisk_chopper/IX_doubledisk_chopper.m:295]: there are some places, where call to `recompute_pdf` was optimized. This is probably one of these places. 

There was a bug in some class, where `check_combo_arg` was returning `obj` but was asiged to `pdf_`. This bug is fixed.

File: _test/common_functions/hashable_obj_tester.m (hunk 1 above)
Comment (abuts) [reply _test/common_functions/hashable_obj_tester.m:9]: hopefully clarified

File: _test/common_functions/hashable_obj_tester.m (hunk 2 above)
Comment (abuts) [reply _test/common_functions/hashable_obj_tester.m:41]: `assertFalse(hobj.hash_defined,sprinft())` would do but this form will print each time `assertFalse` is called in the loop regardless of failure. 
I want to save prining in a loop (old fashoned, I am still remember that it considered very inefficient) so validating `hash_defined` separately. If it fails, error is reported through usual `assert` channel redirected appropriately on every test system.

File: _test/common_functions/hashable_obj_tester.m (hunk 3 above)
Comment (abuts) [reply _test/common_functions/hashable_obj_tester.m:53]: comments added.

File: herbert_core/utilities/classes/@hashable/private/to_hashable_array_.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (cmarooney-stfc) [reply herbert_core/utilities/classes/@hashable/private/to_hashable_array_.m:44]: Alex, these are good points and they should be inserted in comments in the code so future developers know what is going on. Please put your explanations for points we raise as comments in the code

Comment (cmarooney-stfc) [inline herbert_core/utilities/classes/@hashable/private/to_hashable_array_.m:44]: Alex, these are good points and they should be inserted in comments in the code so future developers know what is going on. Please put your explanations for points we raise as comments in the code

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]

File: herbert_core/utilities/classes/@hashable/private/to_hashable_array_.m [hunk 11]
```diff
@@ -32,12 +36,22 @@
...
+            %-----> hash class dependent block
+            % the operation below work only for MD5 hash in string form
+            % if we ever decided to switch to uint64-bit hash, operations
+            % with container modification become an order of magnitude
+            % faster. This block must change is such case.
             if iscell(hash) % use array of hashes as source for final hash
                 hash = strjoin(hash ,'');
             end
-            tm = typecast(hash,'uint8');
+            tm = uint8(hash); % use typecast for numeric hash.
```
Comment (oerc0122) [inline herbert_core/utilities/classes/@hashable/private/to_hashable_array_.m:53]: This comment does not make sense. You aren't using `typecast`

**Review by oerc0122 — CHANGES_REQUESTED** [review]

File: herbert_core/classes/instrument_classes/@Goniometer/Goniometer.m [hunk 12]
```diff
@@ -146,15 +146,19 @@
         %
         function obj = set.omega(obj,val)
             obj.omega_=check_angular_val(obj,val);
+            obj = obj.clear_hash();
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/@Goniometer/Goniometer.m:149]: I thought about this solution but decided its too complicated to be implemented. 

after all, when you design a new class, you carefully work along all properties and be able to check if hashes are appropriately cleared up. 

This solution would also allow you do fine tunning for performance, i.e. checking if properies changed/unchanged and then clear hash or putting some clearing operations into `check_combo_arg` not to perform them multiple times. 

In other words, this solution is much easier to accomodate

As for forgetting to do this change for some classes I am modifying now -- I've added unit test which checks if all hash clearence works fine, and found number of error I did adding these changes.

I am ready to return to that and disuss alternative solution on appropriate ticket.

File: herbert_core/applications/multifit/@mfclass/private/multifit_lsqr_ser.m [hunk 13]
```diff
@@ -458,16 +458,32 @@
 
         [~,s,v]=svd(jac,0);
         % constrain possible conditionality number and nullify small
-        % elements to inprove resulting fit comparison in tests in 
+        % elements to inprove resulting fit comparison in tests in
         % case of poorly defined matrix
         max_sv = max(abs(s(:)));
         degraded = abs(s/max_sv)<eps('single');
+        % see Re #1798 for discussion about meaning of this piece of code.
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/applications/multifit/@mfclass/private/multifit_lsqr_ser.m:465]: this change is necessary to fix unit test and extensively discussed on the meeting. 

It is necessary and the follow up will come

[the remark this answers is not in the document]
Comment (oerc0122) [reply herbert_core/applications/multifit/@mfclass/private/multifit_lsqr_ser.m:465]: Do you mean the meeting where it was stated that it would be a separate PR?

File: _test/test_TF_components/test_tobyfit_refine_moderator.m [hunk 14]
```diff
@@ -33,12 +33,13 @@
             % Initialise test object properties
             obj.w1inc = w1inc;      % Cut over purely incoherent line
             tol_sig = 0.25;     % tolerance as multiple of st. dev. of reference value
-            tol_abs = 0;        % absolute tolerance
-            tol_rel = 0;        % relative tolerance
+            tol_abs = 0.0;        % absolute tolerance
+            tol_rel = 0.02;     % relative tolerance
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_TF_components/test_tobyfit_refine_moderator.m:37]: there was hidden tolerance somewhere in test routines before. Now I pulled it out and made explicit. 

We are talking about fitting tolerance. It is very good tolerance for this purpose and it can not be 0

File: _test/test_instrument_classes/test_IX_detectors_4.m [hunk 15]
```diff
@@ -51,55 +77,55 @@ function test_det_slab (self)
...
         end
-        
+
         %--------------------------------------------------------------------------
-        function test_IX_detector_tobyfitClassic (self)
-            dia = repmat(0.01,9,1);
-            height = repmat(0.02,9,1);
-            det = IX_det_TobyfitClassic(dia, height);
+        function test_IX_detector_tobyfitClassic (~)
+            diaph = repmat(0.01,9,1);
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_instrument_classes/test_IX_detectors_4.m:83]: I suspect its `diaphragm`. As I was not writing the initial  test, I am not sure so write vague reminder to myself

File: _test/test_instrument_classes/test_IX_detectors_4.m [hunk 16]
```diff
@@ -51,55 +77,55 @@ function test_det_slab (self)
...
+        function test_IX_de_He3tube (~)
 
-            height = (0.0001:0.001:0.0099)';
+            hght = (0.0001:0.001:0.0099)';
             wall = (1.8:18:179.999)';
-            atms = (1.8:18:179.999)';
-            dia = wall.*2.2; % dia must be >2*wall, see IX_det_he3tube code
-            tube = IX_det_He3tube(dia, height, wall, atms);
-            assertEqual(tube.dia, dia);
+            atmss = (1.8:18:179.999)';
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_instrument_classes/test_IX_detectors_4.m:114]: have no idea. Not my test.

File: _test/test_mex_nomex/test_main_mex.m [hunk 17]
```diff
@@ -159,38 +159,38 @@ function test_calc_proj_options(obj)
...
             % compare mex and nomex hashes
             assertTrue( strcmp(hash1_nomex, hash1_mex) );
             assertTrue( strcmp(hash2_nomex, hash2_mex) );
-            
+
             % compare hashes against previously recorded values
             assertTrue( strcmp(hash1_nomex, 'b7e33a9818a21b4a33425f56c7751a4a') );
-            assertTrue( strcmp(hash2_nomex, '22e08809d241e24adfb657fc1341ff94') );
-            
+            assertTrue( strcmp(hash2_nomex, 'a656e2a223791fc39723949baebfddbc') );
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_mex_nomex/test_main_mex.m:192]: It has changed because `instrument` become `hashable` and `hashable` class calculates hashes from the values of its `hahshableFields`  and not from serialized array of bytes as before. 

This have been reported on the meeting.

File: _test/test_instrument_classes/test_IX_detectors_4.m (hunk 16 above)
[the remark this answers is not in the document]
Comment (oerc0122) [reply _test/test_instrument_classes/test_IX_detectors_4.m:114]: So why have you changed the name of it?

File: _test/test_instrument_classes/test_IX_detectors_4.m (hunk 15 above)
[the remark this answers is not in the document]
Comment (oerc0122) [reply _test/test_instrument_classes/test_IX_detectors_4.m:83]: I should think it's `diameter`, why would it be `diaphragm`? As a result this is not a helpful change.

File: herbert_core/classes/instrument_classes/instrument/detectors/@IX_detector_array/IX_detector_array.m (hunk 8 above)
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/detectors/@IX_detector_array/IX_detector_array.m:563]: Do not understand this statement. By default,`hashableFields == serializabeFields`. This may be changed in special circumstances. I.e. `detectors_array` have filename and this should be excluded from `hashableFields`. 

neither `sef.filename` should have `clear_hash` function -- one more point of not overloadins `subsassign` globally.

[the remark this answers is not in the document]
Comment (oerc0122) [reply herbert_core/classes/instrument_classes/instrument/detectors/@IX_detector_array/IX_detector_array.m:563]: By making `check_combo_arg` clear the hash, you are forcing any new properties added which require internal consistency to invalidate the hash, even if they aren't a hashed property. 

By doing this rather than (as has been done in other objects) consistently clearing as part of the setter, it will lead to future confusion when trying to extend the object,  

You're changing paradigms in different classes.

File: herbert_core/classes/instrument_classes/instrument/detectors/@IX_det_He3tube/IX_det_He3tube.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/detectors/@IX_det_He3tube/IX_det_He3tube.m:202]: because of 

If you call `check_combo_arg` some or all of your `hashable` fields have changed. Unless you set up all the same values for all properties. Well. Bad luck. By default you need to recalculate hash in this case too. 

If it strongly affects performance -- you are welcome to take measures. There are couple of situations in insturment components where these measures were taken for PDF calculations. hash will benifit from these changes.

as `check_combo_arg` returns `obj`, it may do much more then just checking. and sometimes it does. May be the name is not fully adequate. Well, when it was invented we though it will only check. Not any more.

File: _test/test_instrument_classes/test_IX_detectors_4.m [hunk 18]
```diff
@@ -7,38 +7,64 @@
...
+
         %--------------------------------------------------------------------------
-        function test_det_slab (self)
+        function test_det_slab (~)
             % construct a detector bank with a default IX_det_slab
             % and check its initialisation
             depth = repmat(0.1,9,1);
             width = repmat(0.2,9,1);
-            height = repmat(0.3,9,1);
+            hght = repmat(0.3,9,1);
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_instrument_classes/test_IX_detectors_4.m:50]: Matlab Helper highlights height but ignores width.

File: _test/test_instrument_classes/test_IX_detectors_4.m (hunk 15 above)
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_instrument_classes/test_IX_detectors_4.m:83]: may be its indedd diameter! For some reason I thought about diaphragm. ok, reverted.

File: _test/test_instrument_classes/test_IX_detectors_4.m (hunk 16 above)
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_instrument_classes/test_IX_detectors_4.m:114]: actually this is probably pressure in atmosheres. I've changed the name because Matlab helper complained about it and less red lines in the code -- more chances to sopt something important

File: _test/test_multifit/test_multifit_horace_1.m [hunk 19]
```diff
@@ -115,17 +116,25 @@
 
             % Simulate at the initial parameter values
             wsim_1 = mss.simulate();
-            %c2s = cut(wsim_1,[],[],[-0.1,0.1],[100,120]);
+            % Re #1791 These two statements below should be fixed in #1791
+            % as something wrong in container desing here.
+            disp(wsim_1.detpar.unique_objects.unique_objects)
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_multifit/test_multifit_horace_1.m:121]: Have not noticed that. This fixed by Chris PR #1801 too. Removed.

File: _test/test_sqw_file/test_serialize_deserialize_binary_sqw.m [hunk 20]
```diff
@@ -321,7 +321,7 @@ function test_instrument_conversion(~)
             inst1 = maps_instrument(300,700,'S');
             inst = inst1.to_struct();
             [struc_pos,pos] = ser.calculate_positions(test_format,inst);
-            assertEqual(pos-1,2834);
+            assertEqual(pos-1,3294); % was 2834 before adding hashing field to instrument component
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_sqw_file/test_serialize_deserialize_binary_sqw.m:324]: well, it was such information here (or in other test before) It would certainly not be harmful.

File: _test/test_sqw_pageOpMethods/test_join.m [hunk 21]
```diff
@@ -202,7 +195,7 @@ function test_join_creates_tmp_filebacked_on_conditions(obj)
             % to compare filebacked and memory backed object properly, here
             % we need to have compatible page sizes. The comparison will
             % fail otherwise. Re #1147 -- should fix that.
-            clear clConf;
+            clear clConf; % this makes page sizes compatible
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_sqw_pageOpMethods/test_join.m:198]: The explanation is in previous rows, but I've repeated it in different words here too.

File: _test/test_unique_objects_container/test_unique_references.m [hunk 22]
```diff
@@ -28,20 +28,96 @@ function delete(~)
             unique_references_container.global_container( ...
                 'CLEAR','GLOBAL_NAME_SAMPLES_CONTAINER')
         end
+        %------------------------------------------------------------------
+        function test_save_restore_global_state_two_urcs(~)
+            clWa = set_temporary_warning('off','HERBERT:unique_references_container:debug_only_argument');
+            function urc_clearer()
+                unique_references_container.global_container('CLEAR','multifit_issue')
+            end
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_unique_objects_container/test_unique_references.m:36]: because only 3 tests from couple of dozens use it. Initially it was one. May be worth moving it to obj indeed. Will try it after Cris merges his code, as this is his part and I do not want to deal with unnecessary merge conflicts.

File: herbert_core/applications/multifit/@mfclass/private/multifit_lsqr_ser.m (hunk 13 above)
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/applications/multifit/@mfclass/private/multifit_lsqr_ser.m:465]: yes. And we made ticket about it and will be working on solving it. It is patch here -- it will be either proper solution or it will go and tests would need to be fixed in different way.

File: herbert_core/classes/instrument_classes/@Goniometer/Goniometer.m (hunk 12 above)
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/@Goniometer/Goniometer.m:149]: Above I provided some examples on useful flexibility this design have. 

Again -- not standing on it hard so open for discussion & cost/benifit of implementing other decision. 

If decided, it would not be difficult to remove using global search.

File: herbert_core/utilities/classes/@hashable/private/to_hashable_array_.m (hunk 11 above)
Comment (abuts) [reply herbert_core/utilities/classes/@hashable/private/to_hashable_array_.m:53]: well, if you read sentences 45-49, you will see the context. I have clarified it here too.

File: _test/test_instrument_classes/test_IX_detectors_4.m (hunk 16 above)
[the remark this answers is not in the document]
Comment (oerc0122) [reply _test/test_instrument_classes/test_IX_detectors_4.m:114]: 1.8 to 180 in steps of 18 doesn't sound like pressure in atmospheres to me. Sounds like an angle.

File: herbert_core/applications/multifit/@mfclass/private/multifit_lsqr_ser.m (hunk 13 above)
[the remark this answers is not in the document]
Comment (oerc0122) [reply herbert_core/applications/multifit/@mfclass/private/multifit_lsqr_ser.m:465]: You are asking to merge this into `master`.  Not just experimenting.

[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/applications/multifit/@mfclass/private/multifit_lsqr_ser.m:465]: well, it fixes the test and have physical meaning. It aslo have nothing to do with changes I did and everything to do with poor reliability of fitting procedure which apparently converged to one state in master and another state here. This problem is not to solving here but for dealing within another ticket.

File: _test/test_instrument_classes/test_IX_detectors_4.m (hunk 16 above)
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_instrument_classes/test_IX_detectors_4.m:114]: it is pressure in atmospheres -- I've read class description

File: herbert_core/classes/instrument_classes/instrument/detectors/@IX_detector_array/IX_detector_array.m (hunk 8 above)
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/detectors/@IX_detector_array/IX_detector_array.m:557]: This particular class calls `check_combo_arg` after every setter. 

This is why here I can call `chlear_hash` in check_combo_arg` as it is certain that it will be called at least once.
For every setter contributing to `hashableFields` if you assign them manually or for each constructor or `loadObj` if setters are involved from them. 

This is what I mean by saying `subset`  is not very good for setting hashes. Such subtle obtimization is not available to it. 

Some classes in current `hashable` set are also validating if properties are changed and do nothing if they are not. 
Not sure such check worth doing for every hashable, but if it is already there, it worth using it.

File: herbert_core/classes/instrument_classes/instrument/detectors/@IX_det_He3tube/IX_det_He3tube.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/classes/instrument_classes/instrument/detectors/@IX_det_He3tube/IX_det_He3tube.m:202]: and see answer above, where the same is expressed in more words.
